csm/core/services/usl.py

```diff
@@ -681,7 +681,7 @@ async def post_register_device(
                                 Log.info('Device registration successful')
                                 break
                             elif response.status != 201:
-                                reason = 'Device registration failed'
+                                reason = 'Lyve Pilot failed to register the device'
```

Comment:
How we are generating returning error message for bucket creation or account creation faled.
For example:--
While registering account I  provide a duplicate account name... account creation will fail...
* if i try same in create S3 account page, i use to get proper error message with duplicate account
* in case of Registration page i get error as Account creation failed. Ideally i should get proper error message why it failed

Reply:
Also, why LyveDrive registration failed? do we get some error message from UDS which can be helpful for user to show.

Reply:
Just to make things clear, we took care of keeping error handling separate for every corner case. Look at `_initialize_lyve_pilot_s3_resources()` for more examples---and existing IAM user is only one among several other cases.

Reply:
@tb-s  @avrnv 
On s3 Account I create an account with name "acc1"
I tried to create one more account with same name
https://10.230.255.10:28100/api/v1/s3_accounts
Error I got:--
{"error_code":12288,"message_id":"EntityAlreadyExists","message":"The request was rejected because it attempted to create an account that already exists.","error_format_args":{"account_name":"acc1"}}


I tried to register to UDX and provided same name again "acc1"
https://10.230.255.10:28100/api/v1/udx_device/registration
Output:--
{"error_code":"4101","message":"Internal error: S3 account creation failed"}

UDX does not provide the reason. But S3 account creation REST API from CSM provides the reason.
I want to show similar message on why it failed

Reply:
Same applies for other creation like bucket, IAM users etc.

Reply:
@kupranay As usual, I believe I did not make it clear from my first message.

> Just to make things clear, we took care of keeping error handling separate for every corner case. Look at _initialize_lyve_pilot_s3_resources() for more examples---and existing IAM user is only one among several other cases.

So what you are asking *is already in place* for most of the S3 resources and other potential errors. With regards to an already existing S3 user:

> When we first implemented _initialize_lyve_pilot_s3_resources(), there was no way to check if an S3 user already existed

We can do that now, of course. But this is not what the original ticket is asking for.

Reply:
![image](https://user-images.githubusercontent.com/66459168/95366999-d47c7980-08f1-11eb-8f27-062c2b26c73d.png)

Reply:
> Also, why LyveDrive registration failed? do we get some error message from UDS which can be helpful for user to show.

Like I've already mentioned in JIRA, UDS doesn't provide any explanation except the HTTP error code. Tadeu's already confirmed this.

Reply:
> UDS doesn't provide any explanation except the HTTP 400 error code

@kupranay @avrnv Yes. They only support 3 status codes at this point. An improvement would be to handle status code 400 in separate and handle it as intended (i.e., in cases where the reg token is invalid). Probably it is not at the moment because it is the only error status code supported.
```
      responses:
        200:
          description: successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/RegistrationPortalInfo'
        201:
          description: >
            operation still in progress. 201 will be returned until the
            registration is completed upon completion a 200 will be returned
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Poll'
          headers:
            Location:
              schema:
                type: string
              description: Polling URL
        400:
          description: invalid regcode supplied
          content: {}
```

Reply:
I updated the patch to carefully relay S3 errors to the user. It required refactoring and leveraging S3 services inside the USL in order to reduce the boiler plate code. The patch is huge now, but more things have been thrown away than added. Changes are applied to registration sequence only, so there are places where S3 plugin is still used direcly, but they could be refactored later. @tb-s , keep that in mind and please, review again.
@kupranay , FYI